synchronization/trigger_test.go

```diff
@@ -26,120 +26,133 @@ func newNopLogger() *nopLogger {
 		harness.Supervise(p)
-		time.Sleep(3 * time.Millisecond)
-		require.Equal(t, 0, x, "expected no ticks")
 		p.Stop()
-		require.Equal(t, 0, x, "expected stop to not block")
-
+		wasStopped.EventuallyToggled("expected ticker.Stop() to have been called")
 	})
 }
 
-func TestPeriodicalTriggerStopOnContextCancelWithStopAction(t *testing.T) {
+func TestPeriodicalTriggerStopsTickerOnContextCancel(t *testing.T) {
 	with.Concurrency(t, func(parent context.Context, harness *with.ConcurrencyHarness) {
 		ctx, cancel := context.WithCancel(parent)
-		ch := make(chan struct{})
-		p := synchronization.NewPeriodicalTrigger(ctx, "a periodical trigger", time.Millisecond*2, harness.Logger, func() {}, func() { close(ch) })
+		ticker := synchronization.NewHookTicker()
+		wasStopped := NewTestFlag(t)
+		ticker.Stop_ = func() { wasStopped.Toggle() }
+		p := synchronization.NewPeriodicalTrigger(ctx, "a periodical trigger", ticker, harness.Logger, func() {}, nil)
+		defer p.Stop()
 		harness.Supervise(p)
 		cancel()
-		_, ok := <-ch
-		require.False(t, ok, "expected trigger stop action to close the channel")
-
+		wasStopped.EventuallyToggled("expected ticker.Stop() to have been called")
 	})
 }
 
-func TestPeriodicalTriggerRunsOnStopAction(t *testing.T) {
-	with.Concurrency(t, func(parent context.Context, harness *with.ConcurrencyHarness) {
-		latch := make(chan struct{})
-		x := 0
-		p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", time.Second, harness.Logger, func() { x++ }, func() {
-			x = 20
-			latch <- struct{}{}
-		})
+func TestPeriodicalTriggerCallsStopHookOnCallToStop(t *testing.T) {
+	with.Concurrency(t, func(ctx context.Context, harness *with.ConcurrencyHarness) {
+		ticker := synchronization.NewHookTicker()
+		wasStopped := NewTestFlag(t)
+		p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", ticker, harness.Logger, func() {}, func() { wasStopped.Toggle() })
 		harness.Supervise(p)
 		p.Stop()
-		<-latch // wait for stop to happen...
-		require.Equal(t, 20, x, "expected x to have the stop value")
-
+		wasStopped.EventuallyToggled("expected ticker.Stop() to have been called")
 	})
 }
 
-func TestPeriodicalTriggerKeepsGoingOnPanic(t *testing.T) {
+func TestPeriodicalTriggerCallsStopHookOnContextCancel(t *testing.T) {
 	with.Concurrency(t, func(parent context.Context, harness *with.ConcurrencyHarness) {
-
-		logger := newNopLogger()
-		var handlerInvocationCount uint32
-		p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", time.Millisecond, logger, func() {
-			atomic.AddUint32(&handlerInvocationCount, 1)
-			panic("we should not see this other than the logs")
-		}, nil)
-
-		requireAtomicGreaterThan(t, &handlerInvocationCount, 1, "expected trigger to have ticked more than once even though it panics")
-
-		p.Stop()
+		ctx, cancel := context.WithCancel(parent)
+		ticker := synchronization.NewHookTicker()
```

Comment:
you could create another implementation local to the test, `NopTrigger`

Reply:
that would be more expressive but I don't think increasing readability for these tests justify the boilerplate. I went with a single `TestTicker`  implementation instead. WDYT?